sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -345,240 +312,56 @@ public Read withMaxReadTime(Duration maxReadTime) {
-
-    UnboundedRabbitMqReader(RabbitMQSource source, RabbitMQCheckpointMark checkpointMark)
-        throws IOException {
-      this.source = source;
-      this.current = null;
-      this.checkpointMark = checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark();
-    }
-
-    @Override
-    public Instant getWatermark() {
-      return checkpointMark.latestTimestamp;
-    }
-
-    @Override
-    public UnboundedSource.CheckpointMark getCheckpointMark() {
-      return checkpointMark;
-    }
-
-    @Override
-    public RabbitMQSource getCurrentSource() {
-      return source;
+    public Read withTimestampPolicyFactory(TimestampPolicyFactory timestampPolicyFactory) {
+      return builder().setTimestampPolicyFactory(timestampPolicyFactory).build();
     }
 
     @Override
-    public byte[] getCurrentRecordId() {
-      if (current == null) {
-        throw new NoSuchElementException();
-      }
-      if (currentRecordId != null) {
-        return currentRecordId;
-      } else {
-        return "".getBytes(StandardCharsets.UTF_8);
-      }
-    }
-
-    @Override
-    public Instant getCurrentTimestamp() {
-      if (currentTimestamp == null) {
-        throw new NoSuchElementException();
-      }
-      return currentTimestamp;
-    }
-
-    @Override
-    public RabbitMqMessage getCurrent() {
-      if (current == null) {
-        throw new NoSuchElementException();
-      }
-      return current;
-    }
+    public PCollection<RabbitMqMessage> expand(PBegin input) {
+      org.apache.beam.sdk.io.Read.Unbounded<RabbitMqMessage> unbounded =
+          org.apache.beam.sdk.io.Read.from(new RabbitMqSource(this));
 
-    @Override
-    public boolean start() throws IOException {
-      try {
-        connectionHandler = new ConnectionHandler(source.spec.uri());
-        connectionHandler.start();
-
-        Channel channel = connectionHandler.getChannel();
-
-        queueName = source.spec.queue();
-        if (source.spec.queueDeclare()) {
-          // declare the queue (if not done by another application)
-          // channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
-          channel.queueDeclare(queueName, false, false, false, null);
-        }
-        if (source.spec.exchange() != null) {
-          if (source.spec.exchangeDeclare()) {
-            channel.exchangeDeclare(source.spec.exchange(), source.spec.exchangeType());
-          }
-          if (queueName == null) {
-            queueName = channel.queueDeclare().getQueue();
-          }
-          channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
-        }
-        checkpointMark.channel = channel;
-        channel.txSelect();
-      } catch (IOException e) {
-        throw e;
-      } catch (Exception e) {
-        throw new IOException(e);
-      }
-      return advance();
-    }
+      PTransform<PBegin, PCollection<RabbitMqMessage>> transform = unbounded;
```

Comment:
imo this would be clearer as 
```java
PTransform<PBegin, PCollection<RabbitMqMessage>> transform = (maxNumRecords() < Long.MAX_VALUE || maxReadTime() != null)
  ? unbounded.withMaxReadTime(maxReadTime()).withMaxNumRecords(maxNumRecords())
  : unbounded;
```